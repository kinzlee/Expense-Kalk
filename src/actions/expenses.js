import uuid from 'uuid';

// ACTION GENERATOR for expense  reducer

export const addExpenses = ({
    description = '',
    messageNote = '',
    amount = 0,
    createdAt = 0
} = {}) => {
    return {
        type : 'ADD_EXPENSES',
        expenses : {
            id: uuid(),
            description,
            messageNote,
            amount,
            createdAt
        }
    }
};

export const removeExpense = ({id} = {}) => {
    return {
        type : 'REMOVE_EXPENSE',
        id
    }
};



export const editExpense = (id, update) => {
    return {
        type : 'EDIT_EXPENSE',
        id,
        update
    }
};