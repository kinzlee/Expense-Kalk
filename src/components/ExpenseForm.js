import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';



export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expenses ? props.expenses.description : '',
            messageNote: props.expenses ? props.expenses.messageNote : '',
            amount: props.expenses ? (props.expenses.amount / 100).toString() : '',
            createdAt: props.expenses ? moment(props.expenses.createdAt) :  moment(),
            calendarFocused: false,
            error: ''
        };
    }
    descriptionChange = (elt) => {
        const description = elt.target.value; 
        this.setState(() => ({description}));
    };
    messageNoteChange = (elt) => {
        const messageNote = elt.target.value;
        this.setState(() => ({messageNote}));
    };
    amountChange =(elt) => {
        const amount = elt.target.value;
        if(!amount || amount.match(/^\d{1, }(\.\d{0, 2})?$/)) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({createdAt}));        }
    };
    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused: focused}));
    };
    onSubmit = (elt) => {
        elt.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(()=>({error: 'please provide a description and amount'}));
        }else {
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) *100,
                createdAt: this.state.createdAt.valueOf(),
                messageNote: this.state.messageNote
            })
        }
    };
    render() {
        
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    <input 
                    type="text"
                    placeholder="description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.descriptionChange}
                    />
                    <input 
                    type="text"
                    placeholder="amount"
                    value={this.state.amount}
                    onChange={this.amountChange}
                    />
                    <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    />
                    <textarea
                    placeholder="why do you want to make this expense"
                    value={this.state.messageNote}
                    onChange={this.messageNoteChange}
                    >

                    </textarea>
                    <button
                    
                    >Add Expense
                    </button>
                </form>
            </div>
        )
    }
}