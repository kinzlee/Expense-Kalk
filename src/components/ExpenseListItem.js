import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpensesListItem = ( { id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
        <h2> {description} </h2>
        </Link>
        <p> 
        {numeral(amount).format('#0,0.00')}
         - 
        {moment(createdAt).format('Do, MMMM YYYY')} 
        </p>
    </div>
);

export default ExpensesListItem;