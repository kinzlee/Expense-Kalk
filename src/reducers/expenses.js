// Expenses Reducer
const defaultExpenseState = [];

export default (state=defaultExpenseState, action) => {
    switch(action.type) {
        
        case  'ADD_EXPENSES':
            return [
            ...state,
            action.expenses
            ];

        case 'REMOVE_EXPENSE' : 
            return state.filter( ({id}) => id !== action.id);
            
        case 'EDIT_EXPENSE' :
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.update
                    }
                }
                    return expense
            });

        default : 
        return state

           
    }
};