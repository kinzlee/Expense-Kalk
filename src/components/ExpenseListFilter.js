import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFilterText, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };
    onFocusChange = (calendarFocused) => {
        this.setState( () => ({ calendarFocused }));
    }
    render() {
        return (
            <div>                                                                                                                                                                           
                <input type="text" value={this.props.filter.text} onChange={(elt) => {
                   this.props.dispatch(setFilterText(elt.target.value));
                }} />
                <select 
                value={this.props.filter.sortBy}
                  onChange={(elt) => {
                    if(elt.target.value === "date"){
                        this.props.dispatch(sortByDate(elt.target.value))
                    }else if(elt.target.value === "amount") {
                        this.props.dispatch(sortByAmount(elt.target.value))
                    }
                }}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                startDate={this.props.filter.startDate}
                startDateId = "{this.props.id}"
                endDate={this.props.filter.endDate}
                endDateId="{this.props.id}"
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToFilter = (state) => {
    return {
        filter: state.filter
    }
}

export default connect(mapStateToFilter)(ExpenseListFilter);