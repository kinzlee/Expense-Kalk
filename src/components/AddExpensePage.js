import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpenses } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return  (
            <div>
                <h1> Add Expense</h1>
                <ExpenseForm 
                onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

// const AddExpensePage = (props) => (
//     <div>
//         <h1> Add Expense</h1>
//         <ExpenseForm 
//         addExpense={
//             (expense) => {
//             // props.dispatch(addExpenses(expense));
//             props.addExpense(expense);
//             props.history.push('/');    
//             }
//         }
//         />
//     </div>
// );

const mapDispatchToProps = () => {
    return {
        addExpense: (expense) => dispatch(addExpenses(expense))
    }
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);