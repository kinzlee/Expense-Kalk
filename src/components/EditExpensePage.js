import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';        
import { removeExpense, editExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expenses) => {
        this.props.editExpense(this.props.expenses.id, expenses);
        this.props.history.push('/'); 
    };
    onClick =() => {
        this.props.removeExpense({id: this.props.expenses.id}) 
        this.props.history.push('/');    
        }
    render() {
        return (
            <div>
                <ExpenseForm 
                expenses={this.props.expenses}
                onSubmit={this.onSubmit}
                /> 
                <button
                onClick={this.onClick}
                > remove
                </button>
            </div>
            )
    }
}

// const EditExpensePage = (props) => {
//     console.log(props);
//     return (
//     <div>
//         <ExpenseForm 
//         expenses={props.expenses}
//         onSubmit={(expenses) => {
//         props.dispatch(editExpense(props.expenses.id, expenses));
//         props.history.push('/');    
//       }}
//         /> 
//         <button
//         onClick={() => {
//         props.dispatch(removeExpense({id: props.expenses.id})) 
//         props.history.push('/');    
//         }}
//         > remove
//         </button>
//     </div>
//     )
// };


const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) =>  dispatch(editExpense(id, expense)),
        removeExpense: (data) => dispatch(removeExpense(data)) 
    }
};


const mapStateToProps = (state, props) => {
    return {
        expenses: state.expenses.find((expenses) => expenses.id === props.match.params.id)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
