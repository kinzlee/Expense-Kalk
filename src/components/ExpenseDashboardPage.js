import React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseListFilter from './ExpenseListFilter'; 

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilter/>
        
        <ExpensesList />
    </div>
);

export default ExpenseDashboardPage;