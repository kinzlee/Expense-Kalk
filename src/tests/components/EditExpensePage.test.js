import React from 'react';
import {shallow} from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixture/expenses';

let editExpense, removeExpense, history, wrapper;

beforeEach(() =>{
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(
    <EditExpensePage 
    editExpense={editExpense} 
    removeExpense={removeExpense} 
    history={history}
    expenses={expenses[1]}
    />);
    // wrapper = shallow(<EditExpensePage editExpense={editExpense} removeExpense={re   moveExpense} history={history}/>)
});

test('should render edit expense page', () => {
expect(wrapper).toMatchSnapshot();  
});

test('should handle edit expense', () => {
wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
expect(history.push).toHaveBeenCalledWith('/');
expect(editExpense).toHaveBeenCalledWith(expenses[1].id, expenses[1]);   
});

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenCalledWith('/');
    expect(removeExpense).toHaveBeenCalledWith({ id: expenses[1].id});   
});