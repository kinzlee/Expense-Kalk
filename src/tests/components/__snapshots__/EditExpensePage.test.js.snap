// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page 1`] = `
<div>
  <ExpenseForm
    expenses={
      Object {
        "amount": 45000,
        "createdAt": 0,
        "description": "phone",
        "id": "2",
        "messageNote": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     remove
  </button>
</div>
`;
